shoot-client/network-connection.sh

```diff
@@ -44,98 +44,84 @@ function configure_tcp() {
-
-    local ip=$(echo $cidr | cut -f1 -d/)
-    local numon=$(echo $cidr | cut -f2 -d/)
-
-    local numoff=$(( 32 - $numon ))
-    while [ "$numon" -ne "0" ]; do
-            start=1${start}
-            numon=$(( $numon - 1 ))
-    done
-    while [ "$numoff" -ne "0" ]; do
-        end=0${end}
-        numoff=$(( $numoff - 1 ))
-    done
-    local bitstring=$start$end
-
-    bitmask=$(echo "obase=16 ; $(( 2#$bitstring )) " | bc | sed 's/.\{2\}/& /g')
-
-    for t in $bitmask ; do
-        str=$str.$((16#$t))
-    done
-
-    echo $str | cut -f2-  -d\.
-}
-
-service_network_address=$(echo $service_network | cut -f1 -d/)
-service_network_netmask=$(CIDR2Netmask $service_network)
+vpn_seed_server="vpn-seed-server"
+dev="tun0"
+if [[ -n "$VPN_SERVER_INDEX" ]]; then
+  vpn_seed_server="vpn-seed-server-$VPN_SERVER_INDEX"
+  dev="tap$VPN_SERVER_INDEX"
+fi
+log "using $vpn_seed_server, dev $dev"
 
-pod_network_address=$(echo $pod_network | cut -f1 -d/)
-pod_network_netmask=$(CIDR2Netmask $pod_network)
+# suffix for vpn client secret directory
+suffix=""
+if [[ $POD_NAME =~ .*-([0-4])$ ]]; then
+  suffix="-${BASH_REMATCH[1]}"
+fi
 
-sed -e "s/\${SERVICE_NETWORK_ADDRESS}/${service_network_address}/" \
-    -e "s/\${SERVICE_NETWORK_NETMASK}/${service_network_netmask}/" \
-    -e "s/\${POD_NETWORK_ADDRESS}/${pod_network_address}/" \
-    -e "s/\${POD_NETWORK_NETMASK}/${pod_network_netmask}/" \
+sed -e "s/\${SUFFIX}/${suffix}/" \
     openvpn.config.template > openvpn.config
 
-if [[ ! -z "$node_network" ]]; then
-  for n in $(echo $node_network |  sed 's/[][]//g' | sed 's/,/ /g')
-  do
-      node_network_address=$(echo $n | cut -f1 -d/)
-      node_network_netmask=$(CIDR2Netmask $n)
-      echo "pull-filter ignore \"route ${node_network_address} ${node_network_netmask}\"" >> openvpn.config
-  done
-fi
-
-echo "pull-filter accept \"route 192.168.123.\"" >> openvpn.config
-echo "pull-filter ignore \"route\"" >> openvpn.config
 echo "pull-filter ignore redirect-gateway" >> openvpn.config
 echo "pull-filter ignore route-ipv6" >> openvpn.config
 echo "pull-filter ignore redirect-gateway-ipv6" >> openvpn.config
 
-# enable forwarding and NAT
-iptables --append FORWARD --in-interface tun0 -j ACCEPT
-iptables --append POSTROUTING --out-interface eth0 --table nat -j MASQUERADE
+echo "port ${openvpn_port}" >> openvpn.config
+if [[ "$is_shoot_client" == "true" ]]; then
+  # use http proxy only for vpn-shoot-client
+  echo "http-proxy ${ENDPOINT} ${openvpn_port}" >> openvpn.config
+  echo "http-proxy-option CUSTOM-HEADER Reversed-VPN ${reversed_vpn_header}" >> openvpn.config
+
+  # enable forwarding and NAT
+  iptables --append FORWARD --in-interface $dev -j ACCEPT
+  iptables --append POSTROUTING --out-interface eth0 --table nat -j MASQUERADE
+else
+  if [[ "$VPN_SERVER_INDEX" == "0" ]]; then
+    # start vpn-path-controller for selecting routing path
+    log "starting vpn-path-controller (logs see /path-controller.log)"
+    /path-controller.sh &
```

Comment:
How do we handle failure of the path-controller? I suppose everything would continue to run if the path-controller just dies, but the path will never be updated.

Reply:
I have added a liveness probe to the container running the path-controller. See new script `path-controller-liveness-probe.sh`. It checks if the log has been updated in the last 30 seconds.